token/services/interop/exchange/transaction.go

```diff
@@ -0,0 +1,212 @@
+*/
+
+package exchange
+
+import (
+	"crypto"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"time"
+
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/core/identity"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/interop/encoding"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/services/ttx"
+	"github.com/pkg/errors"
+)
+
+const (
+	ScriptTypeExchange    = "exchange" // exchange script
+	defaultDeadlineOffset = time.Hour
+)
+
+func WithHash(hash []byte) token.TransferOption {
+	return func(o *token.TransferOptions) error {
+		if o.Attributes == nil {
+			o.Attributes = map[interface{}]interface{}{}
+		}
+		o.Attributes["exchange.hash"] = hash
+		return nil
+	}
+}
+
+func WithHashFunc(hashFunc crypto.Hash) token.TransferOption {
+	return func(o *token.TransferOptions) error {
+		if o.Attributes == nil {
+			o.Attributes = map[interface{}]interface{}{}
+		}
+		o.Attributes["exchange.hashFunc"] = hashFunc
+		return nil
+	}
+}
+
+func WithHashEncoding(encoding encoding.Encoding) token.TransferOption {
+	return func(o *token.TransferOptions) error {
+		if o.Attributes == nil {
+			o.Attributes = map[interface{}]interface{}{}
+		}
+		o.Attributes["exchange.hashEncoding"] = encoding
+		return nil
+	}
+}
+
+func compileTransferOptions(opts ...token.TransferOption) (*token.TransferOptions, error) {
+	txOptions := &token.TransferOptions{}
+	for _, opt := range opts {
+		if err := opt(txOptions); err != nil {
+			return nil, err
+		}
+	}
+	return txOptions, nil
+}
+
+type Transaction struct {
+	*ttx.Transaction
+}
+
+func NewTransaction(sp view.Context, signer view.Identity, opts ...ttx.TxOption) (*Transaction, error) {
+	tx, err := ttx.NewTransaction(sp, signer, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return &Transaction{
+		Transaction: tx,
+	}, nil
+}
+
+func (t *Transaction) Outputs() (*OutputStream, error) {
+	outs, err := t.TokenRequest.Outputs()
+	if err != nil {
+		return nil, err
+	}
+	return NewOutputStream(outs), nil
+}
+
+func (t *Transaction) Exchange(wallet *token.OwnerWallet, sender view.Identity, typ string, value uint64, recipient view.Identity, deadline time.Duration, opts ...token.TransferOption) ([]byte, error) {
+	options, err := compileTransferOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+	if deadline == 0 {
+		deadline = defaultDeadlineOffset
+	}
+	if recipient.IsNone() {
+		return nil, errors.Errorf("must specify a recipient")
+	}
+
+	if sender == nil {
+		sender, err = wallet.GetRecipientIdentity()
+		if err != nil {
+			return nil, errors.WithMessagef(err, "failed getting owner identity")
+		}
+	}
+
+	var hash []byte
+	hashFunc := crypto.SHA256 // default hash function
+	var hashEncoding encoding.Encoding
+	if options.Attributes != nil {
+		boxed, ok := options.Attributes["exchange.hash"]
+		if ok {
+			hash, ok = boxed.([]byte)
```

Comment:
We need  a check in the well-formedness (e.g. length) of the hash.

Reply:
this is done later by the `recipientAsScript` func

Reply:
if that's the case, then it's fine.